api/V1/eventUtil.ts

```diff
@@ -69,22 +74,118 @@ async function uploadEvent(request, response) {
+  options.eventType = ["rpi-power-on", "daytime-power-off", "stop-reported"];
+  options.admin = admin;
+  options.useCreatedDate = false;
+  const result = await models.Event.latestEvents(
+    request.user,
+    query.deviceID,
+    options
+  );
+  const deviceEvents = {};
+  for (const event of result.rows) {
+    if (deviceEvents.hasOwnProperty(event.DeviceId)) {
+      deviceEvents[event.DeviceId].update(event);
+    } else {
+      deviceEvents[event.DeviceId] = new PowerEvents(event);
+    }
+  }
+  for (const id of Object.keys(deviceEvents)) {
+    deviceEvents[id].checkIfStopped();
+  }
+
+  return Object.values(deviceEvents) as PowerEvents[];
+}
+
 const eventAuth = [
   middleware.getDetailSnapshotById(body, "eventDetailId").optional(),
   middleware.isDateArray(
     "dateTimes",
     "List of times event occured is required."
   ),
-  body("description.type").matches(EVENT_TYPE_REGEXP).optional(), 
+  body("description.type").matches(EVENT_TYPE_REGEXP).optional(),
   oneOf(
     [body("eventDetailId").exists(), body("description.type").exists()],
     "Either 'eventDetailId' or 'description.type' must be specified.  Description"
   )
 ];
 
+export class PowerEvents {
+  lastReported: Moment | null;
+  lastStarted: Moment | null;
+  lastStopped: Moment | null;
+  Device: Device | null;
+  hasStopped: boolean;
+  hasAlerted: boolean;
+
+  constructor(event: Event) {
+    this.hasStopped = false;
+    this.lastReported = null;
+    this.lastStarted = null;
+    this.lastStopped = null;
+    this.hasAlerted = false;
+    this.update(event);
+  }
+  update(event: Event) {
+    if (!this.Device && event.Device) {
+      this.Device = event.Device;
+    }
+    const eventDate = moment(event.dateTime);
+    switch (event.EventDetail.type) {
+      case "rpi-power-on":
+        if (this.lastStarted == null || eventDate.isAfter(this.lastStarted)) {
+          this.lastStarted = eventDate;
+        }
+        break;
+      case "daytime-power-off":
+        if (this.lastStopped == null || eventDate.isAfter(this.lastStopped)) {
+          this.lastStopped = eventDate;
+        }
+        break;
+      case "stop-reported":
+        if (this.lastReported == null || eventDate.isAfter(this.lastReported)) {
+          this.lastReported = eventDate;
+        }
+    }
+  }
+  checkIfStopped(): boolean {
+    if (this.lastStarted == null) {
+      this.hasStopped = false;
+      return this.hasStopped;
+    }
+
+    // check that started hasn't occured after stopped
+    if (this.lastStopped == null) {
+      // check that the started event was atleast 12 hours ago
+      this.hasStopped = moment().diff(this.lastStarted, "hours") > 12;
+    } else if (this.lastStarted.isAfter(this.lastStopped)) {
+      //check we are atleast 30 minutes after expected stopped time (based of yesterdays)
+      this.hasStopped = moment().diff(this.lastStopped, "minutes") > 24.5 * 60;
+    } else {
+      // check this isn't yesterdays start stop events
```

Comment:
Great this is much clearer.  Though this comment confused me.   Maybe "check that it started in the last 24 hours might be better".

Reply:
yep